dipy/reconst/mapmri.py

```diff
@@ -0,0 +1,666 @@
+               and estimation for three-dimensional q-space mri", ISMRM 2009.
+
+        Examples
+        --------
+        In this example, where the data, gradient table and sphere tessellation
+        used for reconstruction are provided, we model the diffusion signal
+        with respect to the MAPMRI model and compute the real and analytical
+        ODF.
+
+        >>> from dipy.core.gradients import gradient_table
+        >>> from dipy.data import dsi_voxels, get_sphere
+        >>> data, gtab = dsi_voxels()
+        >>> sphere = get_sphere('symmetric724')
+        >>> from dipy.sims.voxel import SticksAndBall
+        >>> data, golden_directions = SticksAndBall(gtab, d=0.0015, S0=1, angles=[(0, 0), (90, 0)], fractions=[50, 50], snr=None)
+        >>> from dipy.reconst.mapmri import MapmriModel
+        >>> radial_order = 4
+        >>> map_model = MapmriModel(gtab, radial_order=radial_order)
+        >>> mapfit = map_model.fit(data)
+        >>> odf= mapfit.odf(sphere)
+        """
+
+        self.bvals = gtab.bvals
+        self.bvecs = gtab.bvecs
+        self.gtab = gtab
+        self.radial_order = radial_order
+        self.lambd = lambd
+        self.eap_cons = eap_cons
+        self.anisotropic_scaling = anisotropic_scaling
+        if (gtab.big_delta is None) or (gtab.small_delta is None):
+            self.tau = 1 / (4 * np.pi ** 2)
+        else:
+            self.tau = gtab.big_delta - gtab.small_delta / 3.0
+        self.eigenvalue_threshold = eigenvalue_threshold
+
+        self.ind = self.gtab.bvals <= bmax_threshold
+        gtab_dti = gradient_table(
+            self.gtab.bvals[self.ind], self.gtab.bvecs[self.ind, :])
+        self.tenmodel = dti.TensorModel(gtab_dti)
+        self.ind_mat = mapmri_index_matrix(self.radial_order)
+        self.Bm = Bmat(self.ind_mat)
+
+    @multi_voxel_fit
+    def fit(self, data):
+
+        tenfit = self.tenmodel.fit(data[..., self.ind])
+
+        evals = tenfit.evals
+        R = tenfit.evecs
+
+        ind_evals = np.argsort(evals)[::-1]
+        evals = evals[ind_evals]
+        R = R[ind_evals, :]
+
+        evals = np.clip(evals, self.eigenvalue_threshold, evals.max())
+
+        if self.anisotropic_scaling:
+            mu = np.sqrt(evals * 2 * self.tau)
+
+        else:
+            mumean = np.sqrt(evals.mean() * 2 * self.tau)
+            mu = np.array([mumean, mumean, mumean])
+
+        qvals = np.sqrt(self.gtab.bvals / self.tau) / (2 * np.pi)
+        qvecs = np.dot(self.gtab.bvecs, R)
+        q = qvecs * qvals[:, None]
+        M = mapmri_phi_matrix(self.radial_order, mu, q.T)
+
+        # This is a simple empirical regularization, to be replaced
+        I = np.diag(self.ind_mat.sum(1)**2)
+
+        if self.eap_cons:
+            if not have_cvxopt:
+                raise ValueError(
+                    'CVXOPT package needed to enforce constraints')
+            import cvxopt.solvers
```

Comment:
I would put this import on top of the file and not inside the function. Because this fit will run for every voxel there is no reason to import the module each time.